### PR TITLE
[SAK-29389] Added option to prepend DOCTYPE to html

### PR DESCRIPTION
**Pull request by adrianfish**
Automatic now prepends the DOCTYPE, and you can explicitly exclude it
,if you wish, with extra options.

### CODE REVIEW

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/HtmlPageFilter.java [hunk 1]
```diff
@@ -59,6 +59,8 @@
 "    <style>body '{ padding: 5px !important; }'</style>\n" +
 "  </head>\n" +
 "  <body>\n";
+
+    private String doctype = "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\" \"http://www.w3.org/TR/html4/loose.dtd\">\n";
```
Comment (ottenhoff) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/HtmlPageFilter.java]: This is a very dated DOCTYPE.

http://www.w3schools.com/tags/tag_doctype.asp

**Comment by ottenhoff** [comment]
My main question: why is there an option for including DOCTYPE?  Why not always include it?

**Comment by buckett** [comment]
Locally we had some HTML pages that only rendered correctly when in quirks
mode, it you always include a DOCTYPE these pages would break.

So we allowed people to revert back to quirks mode to save them from having
to fix HTML.

On 21 May 2015 at 15:14, Sam Ottenhoff <notifications@github.com> wrote:

> My main question: why is there an option for including DOCTYPE? Why not
> always include it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sakaiproject/sakai/pull/555#issuecomment-104291110>.
>



-- 
  Matthew Buckett

**Comment by adrianfish** [comment]
How about using the strict one, rather than transitional.

From http://www.w3.org/wiki/Choosing_the_right_doctype_for_your_HTML_documents

<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">

So there will be a choice between strict and quirky. The other alternative is the HTML5 recommendation of just:

<!DOCTYPE html>

**Comment by ottenhoff** [comment]
> Locally we had some HTML pages that only rendered correctly when in quirks
> mode, it you always include a DOCTYPE these pages would break.

It seems crazy to me to add *another* option to Sakai Resources to deal with some strange instructor content.

The HTML5 doctype is the preferred path forward:

<!DOCTYPE html>

**Comment by adrianfish** [comment]
People often export stuff like word docs into web pages and just c&p that in, so allowing the browser to just have a stab at it is probably a good idea.

**Comment by adrianfish** [comment]
What happens if the markup is quirky?  Does the simple html doctype trigger
strict parsing, or does the browser go easy?

**Comment by ottenhoff** [comment]
Are we mainly referring to IE quirks mode?  What if you modify your front-end load balancer to inject a X-UA-Compatible header on /access/content/ paths?

**Comment by adrianfish** [comment]
I need to get this moving again ...

How about we just settle on two options in the dropdown, auto and never? 'auto' is the default and properly wraps the content in an html document with a doctype if it's html, 'never' just leaves the page as a fragment, as now.

Regarding the doctype: this link ...

http://www.w3.org/wiki/Doctypes_and_markup_styles

reflects what @ottenhoff says, that <!DOCTYPE html> is the preferred way forward, so I'm going to rejig the commit along those lines.

So, how about the two options?

**Comment by ottenhoff** [comment]
I'm an easy +1 on modifying the DOCTYPE.

I'm -1 on making this a feature that is presented to instructors. My reasons:

a) The majority of instructors have no idea what a DOCTYPE is

b) This feature is centered on an IE browser hack

c) This feature is aimed at an incredibly rare use case.  I have seen a need for this hack once of twice in many years of supporting Sakai

**Comment by adrianfish** [comment]
So ...

+1 on adding the doctype
+1 on two items in the dropbox, 'auto' and 'never'?

In other words, no options with the text 'quirks' or anything like that?

**Comment by steveswinsburg** [comment]
I'm not sure we need a dropdown here at all. As Sam says, why does the instructor need to choose this?

**Comment by adrianfish** [comment]
I'm not sure either. If you select 'never' the fragment of html is just left as a fragment, not a valid html document at all, potentially. What's the point of that? The functionality is to create an html page - perhaps we should be validating the document as such and just rejecting it if it doesn't come up to scratch?

**Comment by steveswinsburg** [comment]
Yep, IMO add the doctype in automatically, forget the selection

**Comment by buckett** [comment]
- Originally Sakai didn't top and tail HTML fragments created by resources.
- When we added the feature to top and tail we left in an option to still have it rendered without the top/tail incase people were loading it in that way (eg loading HTML files from resources my AJAX and placing them in the page).
- When we wanted to change the doctype locally added the new one so people could revert back to the old doctype so that they could fix any page rendering problems easiliy.
- Yes the doctype is old, and could be updated but there maybe changes in rendering behaviour although there is less of a jump.
- This is not something I expect 98% of people to use, it was just an escape hatch for when problems were encountered.
- You can check to see what model a page is using with `document.compatMode` and there is more background http://en.wikipedia.org/wiki/Quirks_mode

**Comment by buckett** [comment]
How about we control the options that are presented with config?
```
resources.filter.show.quirks=false
resources.filter.show.never=false
...
```
Then the default can be to disable some of these.

I think locally we have lots of people writing HTML pages and storing them in resources and have tried hard not to break things for them.

**Comment by adrianfish** [comment]
Okay, that sounds like a good compromise. The default will be to show no options and just wrap and add the html5 doctype. I'll add config options to turn on select and the options within.

**Comment by buckett** [comment]
Adrian's updated the pull request with new code.

**Pull request merged**